src/what-unsafe-does.md

```diff
@@ -54,18 +53,97 @@ empty, "dangling" is the same as "non-null". Note that slices point to their
+Padding bytes aren't technically part of a value's memory, and so may be left
+uninitialized. For unions, this includes the padding bytes of *all* variants,
+as unlike enums, unions are never definitely set to any particular variant (Rust
+does not have the C++ notion of an "active member"). This makes unions
+are the preferred mechanism for working directly with uninitialized memory (see
+[MaybeUninit][] for details).
+
+In simple and blunt terms: you cannot ever even *suggest* the existence of an
+invalid value. No, it's not ok if you "don't use" or "don't read" the value.
+Invalid values are **instant Undefined Behaviour**. The only correct way to
+manipulate memory that could be invalid is with raw pointers using methods
+like write and copy. If you want to leave a local variable or struct field
+uninitialized (or otherwise invalid), you must use a union (like MaybeUninit)
+or enum (like Option) which clearly indicates at the type level that this
+memory may not be part of any value.
+
+
+
+
+## Other Sources of Undefined Behavior
+
+That's it. That's all the causes of Undefined Behavior baked into Rust.
+
+Well, ok, only sort of.
+
+While it's true that the language itself doesn't define that much Undefined
+Behavior, libraries may use unsafe functions and unsafe traits to define
+their own contracts with Undefined Behavior at stake. For instance, the raw
+allocator APIs declare that you aren't allowed to deallocate unallocated memory,
+and the Send trait declares that implementors must in fact be safe to move to
+another thread.
+
+Usually these constraints are in place because violating them will lead to one
+of Rust's Fundamental Undefined Behaviors, but that doesn't have to be the case.
+In particular, several standard library APIs are actually thin wrappers around
+*intrinsics* which tell the compiler it can make certain assumptions.
+
+It's useful to distinguish between these "intrinsic" sources of UB and
+the fundamental ones because the intrinsic ones *don't matter* unless someone
+actually invokes the relevant functions. The fundamental ones, on the other hand,
+are ever-present.
+
+With that said, some intrinsics, like the surprisingly strict [`ptr::offset`][],
+are *pretty* close to fundamental. 😅
+
+
+
+## Not Technically Fundamental Undefined Behavior
+
+There are a few things in Rust that aren't *technically* Fundamental Undefined Behavior,
+but which library authors can implicitly assume don't happen, with Undefined
+Behavior at stake. As such, it should be impossible to do these things in safe
+code, as they can very easily lead to Undefined Behavior.
+
+This section is non-exhaustive, although that may change in the future.
+
+It is *technically not* Undefined Behavior to run a value's destructor twice.
+Authors of destructors may however assume this doesn't happen. For instance, if
+you drop a Box twice it will almost certainly result in Undefined Behavior.
```

Comment:
No need to hedge, this is a double-free.

Reply:
I am hedging here because technically I think it's not UB if someone manages to reallocate the pointer before you run the second drop. You just freed their allocation, and they're probably going to Do An UB, but not necessarily. Or is their some fancy compiler-knows-about-malloc shenanigans where this is still UB because the compiler "knows" you're not allowed to get lucky and free someone else[s allocation like that?

Reply:
Pointer provenance rules (even the basic C-style provenance described in [the UCG glossary](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#pointer-provenance)) mean that *even if* the physical address gets reused, you would not be allowed to use old pointers to access the new allocation.

Reply:
sweet